presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java

```diff
@@ -152,6 +160,91 @@ public static TupleDomainFilter toFilter(Domain domain, int notInThreshold)
         return MultiRange.of(rangeFilters, nullAllowed, false);
     }
 
+    /**
+     * Returns the exclusive number list if it could use NOT IN filter optimization.
+     *
+     * We're not able to distinguish NOT IN statement from normal range compare
+     * E.g.
+     * NOT IN (9, 10, 25) would be [Long.MIN_VALUE, 8] [11, 24] [26, Long.MAX_VALUE]
+     * (num < 8 OR (num > 11 AND number < 24) OR num > 26) could be the same
+     *
+     * When count of exclusive numbers is under the threshold, we can use NOT IN optimization
+     * it would return exclusive number list; otherwise empty list.
+     */
+    private static Optional<List<Long>> getExclusives(List<BigintRange> bigintRanges, int notInThreshold)
+    {
+        if (bigintRanges.size() < 2) {
+            return Optional.empty();
+        }
+
+        long count = getExclusiveCount(bigintRanges);
```

Comment:
I'd combine the counting and collecting the numbers into a single routine. This way you can return as soon as count reaches the threshold and avoid duplicated logic. Here is what I have in mind:

```
    private static Optional<List<Long>> getExclusives(List<BigintRange> bigintRanges, int notInThreshold)
    {
        if (bigintRanges.size() < 2) {
            return Optional.empty();
        }

        if (bigintRanges.get(0).getLower() != Long.MIN_VALUE ||
                bigintRanges.get(bigintRanges.size() - 1).getUpper() != Long.MAX_VALUE) {
            return Optional.empty();
        }

        ImmutableList.Builder<Long> exclusiveNumbers = new ImmutableList.Builder<>();
        int count = 0;
        long prevUpper = bigintRanges.get(0).getUpper();
        for (int i = 1; i < bigintRanges.size(); i++) {
            long lower = bigintRanges.get(i).getLower();

            count += lower - prevUpper - 1;
            if (count > notInThreshold) {
                return Optional.empty();
            }

            for (long n = prevUpper + 1; n < lower; n++) {
                exclusiveNumbers.add(n);
            }

            prevUpper = bigintRanges.get(i).getUpper();
        }

        return Optional.of(exclusiveNumbers.build());
    }
```

Reply:
Yes, it is also my first version, the only concern that I had at that time is that our threshold would be big, we need to allocate the memory and copy values to the `exclusiveNumbers` then throw it away when it exceed our threshold. 
And the frequency of it happening would be big given that we need to check `NOT IN`, `Range`, `NOT Equal`  etc, basically everything except `IN` predicate

So I separate them to make the checking step more efficient

Reply:
@mbasmanova I actually think we can do counting first. After counting then we know exactly how many values there're and can just allocate the memory at one shot without using ImmutableList.Builder<Long>. This should be more efficient because ImmutableLIst.add does check size and grow the memory on the fly:
```
@CanIgnoreReturnValue
    @Override
    public Builder<E> add(E element) {
      checkNotNull(element);
      getReadyToExpandTo(size + 1);
      contents[size++] = element;
      return this;
    }
```